// @ijl/cli/src/webpack/base/index.js
const path = require("path");
const { CleanWebpackPlugin } = require("clean-webpack-plugin");
const webpackCopy = require("copy-webpack-plugin");
const { CustomizeRule, mergeWithRules } = require("webpack-merge");

const isProd = process.env.NODE_ENV === 'production';

const getProjectConfig = require("../../server/utils/get-project-config");

const { webpackConfig = {} } = getProjectConfig();

const outputDirectory = "dist";

const cwd = process.cwd();

const result = mergeWithRules({
  module: {
    rules: {
      test: CustomizeRule.Match,
      use: {
        loader: CustomizeRule.Match,
        options: CustomizeRule.Replace
      }
    }
  }
})(
  {
    output: {
      filename: "[name].js",
      path: path.resolve(cwd, outputDirectory),
      libraryTarget: "umd",
      globalObject: `(typeof self !== 'undefined' ? self : this)`
    },
    plugins: [
      new CleanWebpackPlugin(),
      new webpackCopy({
        patterns: [
          { from: "./remote-assets/", to: "remote-assets", noErrorOnMissing: true },
          { from: process.env.LOCALES, to: "locales", noErrorOnMissing: true }
        ],
        
      })
    ],
    resolve: {
      modules: [cwd, "node_modules"],
      extensions: [".webpack.js", ".web.js", ".ts", ".tsx", ".js", ".css"]
    },

    module: {
      rules: [
        {
          test: /\.tsx?$/,
          loader: "ts-loader"
        },
        {
          test: /\.(jpe?g|gif|png|woff|woff2|svg|ttf|eot|wav|mp3)$/,
          loader: "file-loader"
        },
        {
          test: /\.css$/i,
          use: [
            {
              loader: 'style-loader'
            },
            {
              loader: 'css-loader',
              options: {
                  modules: {
                      mode: 'local',
                      exportGlobals: true,
                      localIdentName: isProd ?
                          '[hash:base64]'
                          : '[path]--[name]__[local]--[hash:base64:3]',
                      localIdentContext: path.resolve(__dirname, 'src'),
                      exportLocalsConvention: "camelCase",
                  },
              }
            },
            {
              loader: 'postcss-loader',
              options: {
                postcssOptions: {
                  plugins: [
                      "postcss-preset-env"
                  ],
                },
              },
            },
          ],
        },
      ]
    },
  },
  webpackConfig
);

module.exports = result;
